src/main/java/com/ncc/aif/AIFProgressMonitor.java

```diff
@@ -0,0 +1,108 @@
+package com.ncc.aif;
+
+import ch.qos.logback.classic.Logger;
+import org.topbraid.jenax.progress.NullProgressMonitor;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Progress monitor for AIF validation.
+ * Note that this class is implemented based on the way the TopBraid validator engine sends progress monitor updates.
+ * Writes a tab-delimited file in the following format:
+ * <code>Shape# | Shape Name | Start Time | End Time | Duration (ms)</code>
+ * <code>1 | sh:MinExclusiveConstraintComponent | t1 | t2 | 25ms<code>
+ * <code>2 | sh:NotConstraintComponent | t1 | t2 | 1225ms<code>
+ */
+class AIFProgressMonitor extends NullProgressMonitor {
+    private static final boolean LOGGING = false;
+    private static final Logger logger = (Logger) (org.slf4j.LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME));
+    private static final SimpleDateFormat FORMAT = new SimpleDateFormat("HH:mm:ss.SSS");
+    private static final int LOGGING_THRESHOLD = 1000;
+    private BufferedWriter out;
+    private Map<Integer, String> shapeNameMap = new HashMap<>();
+    private Map<Integer, Long> shapeTimeMap = new HashMap<>();
+    private int shapeNum;
+    private int numShapes;
+
+    AIFProgressMonitor(String name) throws IOException {
+        final String filename = name.replace(".ttl", "-progress.tab");
```

Comment:
Yeah, I guess at the time I was trying to keep the API more like the TopBraid progress monitors, where you just specify a name.  And I was going to keep the fact that it wrote it to a file as an implementation detail (which changed-- initially the PM just logged to StdOut).  Maybe there should be something in the constructor which specifies whether you want output to a file, StdOut, or both.  Or maybe have two constructors-- the default which logs to StdOut, and one which takes a filename which logs to that filename.  `ValidateAIF.setProgressMonitor()` would take a String filename, which if null would use the default constructor that logs to StdOut.

Does that work for you?

Reply:
Yeah, that's fine with me. The formatting will be the same whether it's writing to stdout or a file?

Reply:
Actually not...when to StdOut it's not a table and it's not tab-delimited.  It uses the "logging" capability of the class, instead of file output.  Also, if creating a file-based monitor fails due to i/o, it can revert to writing to stdout.

Reply:
Basically, it writes out each shape as it starts (via `subtask()`) then logs duration and timestamp of anything which takes longer than a certain threshold (which defaults to 1s).  Basically, this is the output behavior prior to adding the tab-delimited file requirement.